lib/remote/jsonrpcconnection.cpp

```diff
@@ -76,20 +78,55 @@ void JsonRpcConnection::HandleIncomingMessages(boost::asio::yield_context yc)
 		}
 
 		m_Seen = Utility::GetTime();
+		if (m_Endpoint) {
+			m_Endpoint->AddMessageReceived(jsonString.GetLength());
+		}
+
+		auto start (ch::steady_clock::now());
+		ch::steady_clock::duration cpuBoundDuration, totalDuration;
+
+		String rpcMethod("UNKNOWN");
+		String diagnosticInfo; // Contains the diagnostic/debug information in case of an error.
 
 		try {
+			Defer extractTotalDuration ([&start, &totalDuration]() {
+				totalDuration = ch::steady_clock::now() - start;
+			});
+
 			CpuBoundWork handleMessage (yc);
 
+			// Cache the elapsed time to acquire a CPU semaphore used to detect extremely heavy workloads.
+			cpuBoundDuration = ch::steady_clock::now() - start;
+
+			Dictionary::Ptr message = JsonRpc::DecodeMessage(jsonString);
+			if (auto method = message->Get("method"); !method.IsEmpty()) {
+				rpcMethod = method;
+			}
+
 			MessageHandler(message);
 
 			l_TaskStats.InsertValue(Utility::GetTime(), 1);
 		} catch (const std::exception& ex) {
-			Log(m_ShuttingDown ? LogDebug : LogWarning, "JsonRpcConnection")
-				<< "Error while processing JSON-RPC message for identity '" << m_Identity
-				<< "': " << DiagnosticInformation(ex);
+			diagnosticInfo = DiagnosticInformation(ex);
+		}
 
-			break;
+		auto severity = LogDebug;
+		if (totalDuration >= ch::seconds(5) || (!m_ShuttingDown && !diagnosticInfo.IsEmpty())) {
+			// Either this RPC message took an unexpectedly long time to process, or a fatal error
+			// has occurred, so promote the log entry from debug to warning.
+			severity = LogWarning;
+		}
+
+		Log statsLog(severity, "JsonRpcConnection");
+		statsLog << (!diagnosticInfo.IsEmpty() ? "Error while processing" : "Processing") << " JSON-RPC '"
+			<< rpcMethod << "' message for identity '" << m_Identity << "' ";
+
+		if (cpuBoundDuration >= ch::seconds(1)) {
+			statsLog << "waited " << ch::duration_cast<ch::milliseconds>(cpuBoundDuration).count() << "ms on semaphore and ";
 		}
+
+		statsLog << "took total " << ch::duration_cast<ch::milliseconds>(totalDuration).count()
```

Comment:
Why? That's exactly how it's done for HTTP connections. Apart from that, this text is part of the actual log message:
```bash
[2024-09-20 13:20:32 +0000] warning/JsonRpcConnection: Processing JSON-RPC 'event::CheckResult' message for identity 'master-2' waited 26491ms and took total 38855ms.
```

Reply:
Because the message looks strange to me in case of an error. Alternatively, you can always start with _Processed_ and optionally append ". Error: ..."

Reply:
> Alternatively, you can always start with _Processed_ and optionally append ". Error: ..."

That's a better idea!